### PR TITLE
Enable AppFramework dispatcher to enforce integer ranges

### PR DESCRIPTION
**Pull request by blizzz**
## Summary

Psalm allows to specify an int range for integer parameters. We can consider them in the Http Dispatcher so that incoming int values are checked for their range directly in the app framework. Example use case for dashboard added.

Set 28 as milestone, but could go into 29 only. 

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [X] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/AppFramework/Utility/ControllerMethodReflector.php [hunk 1]
```diff
@@ -54,26 +55,38 @@ public function reflect($object, string $method) {
...
 				}
 
-				$this->annotations[$annontation] = [$annotationValue];
+				$this->annotations[$annotation] = [$annotationValue];
 			}
 
 			// extract type parameter information
 			preg_match_all('/@param\h+(?P<type>\w+)\h+\$(?P<var>\w+)/', $docs, $matches);
 			$this->types = array_combine($matches['var'], $matches['type']);
+			preg_match_all('/@psalm-param\h+(?P<type>\w+)<(?P<rangeMin>(-?\d+|min)),\h*(?P<rangeMax>(-?\d+|max))>\h+\$(?P<var>\w+)/', $docs, $matches);
```
Comment (provokateurin) [inline lib/private/AppFramework/Utility/ControllerMethodReflector.php:78]: This also has to work on the normal `@param`

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: lib/private/AppFramework/Utility/ControllerMethodReflector.php (hunk 1 above)
Comment (provokateurin) [reply lib/private/AppFramework/Utility/ControllerMethodReflector.php:78]: Well psalm happily parses it. I'm not sure which standard we try to follow

Comment (provokateurin) [reply lib/private/AppFramework/Utility/ControllerMethodReflector.php:78]: But you are right, we should implement it in openapi-extractor to have it actually usable.

Comment (blizzz) [reply lib/private/AppFramework/Utility/ControllerMethodReflector.php:78]: > Well psalm happily parses it. I'm not sure which standard we try to follow

psalm should not hijack it ;)

**Review by st3iny — APPROVED** [review]
Tested and works. Fancy stuff :sparkles:

**Comment by nickvergessen** [comment]
>  ```diff
> -                        "description": "Limit number of result items per widget",
> +                        "description": "Limit number of result items per widget, not more than 30 are allowed",
> ```

Need to rebuild openapi with `composer run openapi`

**Review by nickvergessen — APPROVED** [review]

**Comment by blizzz** [comment]
> > ```diff
> > -                        "description": "Limit number of result items per widget",
> > +                        "description": "Limit number of result items per widget, not more than 30 are allowed",
> > ```
> 
> Need to rebuild openapi with `composer run openapi`

```
$ composer i && composer run openapi
…
Script "openapi" is not defined in this package
```

:panda_face:

**Comment by blizzz** [comment]
```build/openapi-checker.sh``` it shall be

**Comment by blizzz** [comment]
@provokateurin yeah or no?

**Review by provokateurin — APPROVED** [review]
Sorry, I thought this was solved already. I opened an issue in openapi-extractor to handle the `@psalm-*` annotations and there is another one already for supporting these integer ranges. So at some point this will also properly show up in the openapi docs :)

**Comment by provokateurin** [comment]
(Although I think it is debatable if 30 is a good limit since it is still quite high :woman_shrugging:)

**Comment by provokateurin** [comment]
And 0 as lower limit is also not useful, maybe set it to at least 1

**Comment by blizzz** [comment]
Checked by with @st3iny, he also thinks 30 is a good value.

Either way, the numbers are mostly gut estimations, if necessary we can always tune them.

The 0  I increase to 1.

**Pull request merged**